src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -173,13 +167,11 @@ private DecisionDagBuilder(CSharpCompilation compilation, LabelSymbol defaultLab
         {
             var rootIdentifier = BoundDagTemp.ForOriginalInput(switchExpressionInput);
             int i = 0;
-            var builder = ArrayBuilder<StateForCase>.GetInstance(switchArms.Length);
+            using var builder = TemporaryArray<StateForCase>.GetInstance(switchArms.Length);
```

Comment:
testing in roslyn and tests, >75% of all cases are `<=` 4, so a temp array is a good default choice that does everything on the stack for thsi common case, but goes to heap immediately (with ArrayBuilder) in the rarer case where it's more.